Components/Inputs/TLV2544.cs

```diff
@@ -0,0 +1,264 @@
+                case VoltageReference.INTERNAL_2V: return 2;
+                case VoltageReference.INTERNAL_4V: return 4;
+                default: return double.NaN;
+            }
+        }
+
+        /// <summary> Reads an input channel. </summary>
+        /// <param name="Channel"> 0 to 4 for regular input channels, -1 to -3 for test voltages. </param>
+        /// <returns> The raw ADC reading, which needs further processing and referencing in order to be a usable voltage. </returns>
+        private ushort DoInputRead(sbyte Channel)
+        {
+            Command ChSel;
+            switch (Channel)
+            {
+                case 0: ChSel = Command.SEL_CH0; break;
+                case 1: ChSel = Command.SEL_CH1; break;
+                case 2: ChSel = Command.SEL_CH2; break;
+                case 3: ChSel = Command.SEL_CH3; break;
+                case -1: ChSel = Command.SEL_TEST1; break;
+                case -2: ChSel = Command.SEL_TEST2; break;
+                case -3: ChSel = Command.SEL_TEST3; break;
+                default: ChSel = Command.SEL_CH0; break;
+            }
+            if (this.Config.ConversionMode == ConversionMode.SINGLE_SHOT)
+            {
+                ushort StartRead = DoCommand(ChSel, Long: true);
+                return DoCommand(Command.READ_FIFO);
+            }
+            return 0;
+        }
+
+        /// <summary> Does a 12b read/write with the specified command. </summary>
+        /// <param name="Command"> The command (4 MSb) to send. </param>
+        /// <param name="Data"> The data (12 LSb) to send. </param>
+        /// <param name="Long"> Whether to send out additional SCLK pulses for the ADC to do sampling. </param>
+        /// <returns> The 12b data returned by the device. </returns>
+        private ushort DoCommand(Command Command, ushort Data = 0x000, bool Long = false)
+        {
+            byte[] DataOut;
+            if (!Long) { DataOut = new byte[2]; } // Regular command, no sampling.
+            else
+            {
+                switch (this.Config.ConversionClockSrc) // We are doing sampling, determine the correct number of SCLKs.
+                {
+                    case ConversionClockSrc.SCLK:
+                        DataOut = (this.Config.UseLongSample ? new byte[2 + 4] : new byte[2 + 2]);
+                        break;
+                    case ConversionClockSrc.SCLK_HALF:
+                        DataOut = (this.Config.UseLongSample ? new byte[2 + 7] : new byte[2 + 4]);
+                        break;
+                    case ConversionClockSrc.SCLK_QUARTER:
+                        DataOut = (this.Config.UseLongSample ? new byte[2 + 14] : new byte[2 + 7]);
+                        break;
+                    case ConversionClockSrc.INTERNAL: // We are assuming the time between The two SPI transactions is at least 4us for short, or 8us for long sampling. This seems to always be true on the Pi, but is not guaranteed.
```

Comment:
This comment might be important enough to warrant going in the method comment or class comment, even though I doubt people will have problems with this.

Reply:
Done.